src/store/index-level.ts

```
@@ -1,213 +1,435 @@
+   *
+   * @param matchFilters the filters passed to the parent query.
+   * @param searchFilters the modified filters used for the LevelDB query to search for a subset of items to match against.
+   */
+  async filteredIndexQuery(
+    tenant: string,
+    matchFilters: Filter[],
+    searchFilters:Filter[],
+    queryOptions: QueryOptions,
+    options?: IndexLevelOptions
+  ): Promise<string[]> {
+    const { sortProperty, sortDirection = SortDirection.Ascending, cursor, limit } = queryOptions;
+
+    // we create a matches map so that we can short-circuit matched items within the async single query below.
+    const matches:Map<string, IndexedItem> = new Map();
+
+    try {
+      await Promise.all(searchFilters.map(filter => {
+        return this.executeSingleFilterQuery(tenant, filter, matchFilters, sortProperty, matches, options );
+      }));
+    } catch (error) {
+      if ((error as DwnError).code === DwnErrorCode.IndexInvalidSortProperty) {
+        // return empty results if the sort property is invalid.
+        return [];
+      }
+    }
+
+    const sortedValues = [...matches.values()].sort((a,b) => this.sortItems(a,b, sortProperty, sortDirection));
 
-    // delete all indexes associated with the data of the given ID
-    const ops: LevelWrapperBatchOperation<string>[] = [ ];
-    for (const propertyName in indexes) {
-      const propertyValue = indexes[propertyName];
-      const key = this.join(propertyName, this.encodeValue(propertyValue), dataId);
-      ops.push({ type: 'del', key });
+    // we find the cursor point and only return the result starting there + the limit.
+    // if there is no cursor index, we just start in the beginning.
+    const cursorIndex = cursor ? sortedValues.findIndex(match => match.itemId === cursor) : -1;
+    if (cursor !== undefined && cursorIndex === -1) {
+      // if a cursor is provided but we cannot find it, we return an empty result set
+      return [];
```

Should we return an error instead?

---

Added https://github.com/TBD54566975/dwn-sdk-js/issues/633 to track.